File: services/api-gateway/src/routes/exercise.ts
/**
 * 练习题路由 — 转发到 AI Service 生成多邻国式练习题
 */

import { Router, Request, Response } from 'express';
import { aiServiceClient } from '@/clients/aiServiceClient';
import { logger } from '@/utils/logger';
import { quotaCheck } from '@/middleware/quotaMiddleware';
import { requireFeature } from '@/middleware/requireFeature';
import { recordUsage } from '@/services/usageService';

const router = Router();

/**
 * 生成练习题
 * POST /api/exercise/generate
 */
router.post('/generate', requireFeature('aiChat'), quotaCheck, async (req: Request, res: Response): Promise<void> => {
  try {
    const { exerciseType, topic, difficulty, count } = req.body;

    if (!exerciseType) {
      res.status(400).json({ success: false, error: '缺少 exerciseType 参数' });
      return;
    }

    const result = await aiServiceClient.generateExercise({
      exerciseType,
      topic,
      difficulty,
      count,
    });

    // 记录 AI 用量
    const userId = req.user?.id;
    const totalTokens = (result as any)?.token_usage?.total_tokens;
    if (userId && totalTokens && totalTokens > 0) {
      recordUsage(userId, 'ai', totalTokens).catch(err => {
        logger.error('练习题生成用量记录失败:', err);
      });
    }

    res.json({ success: true, data: result });
  } catch (error: unknown) {
    const axiosErr = error as any;
    const status = axiosErr.response?.status;
    if (status === 400) {
      res.status(400).json({ success: false, error: axiosErr.response?.data?.detail || '参数错误' });
      return;
    }
    logger.error('生成练习题失败:', error);
    res.status(500).json({ success: false, error: '生成练习题失败' });
  }
});

/**
 * 解释答案
 * POST /api/exercise/explain
 */
router.post('/explain', requireFeature('aiChat'), quotaCheck, async (req: Request, res: Response): Promise<void> => {
  try {
    const { question, isCorrect } = req.body;

    if (!question) {
      res.status(400).json({ success: false, error: '缺少 question 参数' });
      return;
    }

    const result = await aiServiceClient.explainExercise({ question, isCorrect });

    // 记录 AI 用量
    const userId = req.user?.id;
    const totalTokens = (result as any)?.token_usage?.total_tokens;
    if (userId && totalTokens && totalTokens > 0) {
      recordUsage(userId, 'ai', totalTokens).catch(err => {
        logger.error('练习题解释用量记录失败:', err);
      });
    }

    res.json({ success: true, data: result });
  } catch (error: unknown) {
    logger.error('解释练习题失败:', error);
    res.status(500).json({ success: true, data: { explanation: '解释生成失败，请稍后重试' } });
  }
});

/**
 * 冒险场景对话
 * POST /api/exercise/adventure/respond
 */
router.post('/adventure/respond', requireFeature('aiChat'), quotaCheck, async (req: Request, res: Response): Promise<void> => {
  try {
    const { scenarioTitle, character, objectives, conversationHistory } = req.body;

    if (!scenarioTitle || !character) {
      res.status(400).json({ success: false, error: '缺少必要参数' });
      return;
    }

    const result = await aiServiceClient.adventureRespond({
      scenarioTitle,
      character,
      objectives,
      conversationHistory,
    });

    // 记录 AI 用量
    const userId = req.user?.id;
    const totalTokens = (result as any)?.token_usage?.total_tokens;
    if (userId && totalTokens && totalTokens > 0) {
      recordUsage(userId, 'ai', totalTokens).catch(err => {
        logger.error('冒险对话用量记录失败:', err);
      });
    }

    res.json({ success: true, data: result });
  } catch (error: unknown) {
    logger.error('冒险场景对话失败:', error);
    // 返回兜底回复
    res.json({
      success: true,
      data: {
        response: "I'm sorry, could you say that again?",
        completedObjectives: [],
      },
    });
  }
});

export default router;

File: services/ai-service/src/routers/exercise.py
"""练习题生成 Router"""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from services.exercise_agent import generate_exercise, explain_exercise, adventure_respond
from utils.logger import logger

router = APIRouter(prefix="/exercise", tags=["exercise"])


class ExerciseRequest(BaseModel):
    exercise_type: str
    topic: str = "daily life"
    difficulty: str = "medium"
    count: int = 1


class ExerciseResponse(BaseModel):
    questions: List[Dict[str, Any]]


class ExplainRequest(BaseModel):
    question: Dict[str, Any]
    is_correct: bool = False


class ExplainResponse(BaseModel):
    explanation: str


class AdventureRequest(BaseModel):
    scenario_title: str
    character: str
    objectives: List[str] = []
    conversation_history: List[Dict[str, str]] = []


class AdventureResponse(BaseModel):
    response: str
    completedObjectives: List[int] = []


@router.post("/generate", response_model=ExerciseResponse)
async def generate(request: ExerciseRequest):
    """生成练习题"""
    try:
        result = await generate_exercise(
            exercise_type=request.exercise_type,
            topic=request.topic,
            difficulty=request.difficulty,
            count=request.count,
        )
        return result
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"生成练习题失败: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/explain", response_model=ExplainResponse)
async def explain(request: ExplainRequest):
    """解释练习题答案"""
    try:
        result = await explain_exercise(
            question=request.question,
            is_correct=request.is_correct,
        )
        return result
    except Exception as e:
        logger.error(f"解释练习题失败: {e}")
        return ExplainResponse(explanation="解释生成失败，请稍后重试")


@router.post("/adventure/respond", response_model=AdventureResponse)
async def adventure(request: AdventureRequest):
    """冒险场景对话"""
    try:
        result = await adventure_respond(
            scenario_title=request.scenario_title,
            character=request.character,
            objectives=request.objectives,
            conversation_history=request.conversation_history,
        )
        return result
    except Exception as e:
        logger.error(f"冒险场景对话失败: {e}")
        return AdventureResponse(
            response="I'm sorry, could you say that again?",
            completedObjectives=[],
        )
